cppcache/src/ExecutionImpl.cpp

```diff
@@ -329,26 +281,23 @@ std::shared_ptr<ResultCollector> ExecutionImpl::execute(
           "Execution::execute: Transaction function execution on pool is not "
           "supported");
     }
-    if (m_allServer == false) {
-      executeOnPool(
-          func, isHAHasResultOptimizeForWrite,
-          (isHAHasResultOptimizeForWrite & 1) ? m_pool->getRetryAttempts() : 0,
-          timeout);
-      if (serverHasResult == true) {
-        // ExecutionImpl::addResults(m_rc, rs);
+    if (!m_allServer) {
+      executeOnPool(func, attrs, attrs.isHA() ? m_pool->getRetryAttempts() : 0,
+                    timeout);
+      if (attrs.hasResult()) {
         m_rc->endResults();
       }
       return m_rc;
     }
-    executeOnAllServers(func, isHAHasResultOptimizeForWrite, timeout);
+    executeOnAllServers(func, attrs, timeout);
   } else {
     throw IllegalStateException("Execution::execute: should not be here");
   }
   return m_rc;
 }
 
-GfErrType ExecutionImpl::getFuncAttributes(
-    const std::string& func, std::shared_ptr<std::vector<int8_t>>* attr) {
+GfErrType ExecutionImpl::getFuncAttributes(const std::string& func,
+                                           FunctionAttributes& attr) {
```

Comment:
Does it make more sense to have the function return `FunctionAttributes` and `throw` exceptions?